src/openassetio-core/src/pluginSystem/HybridPluginSystemManagerImplementationFactory.cpp

```diff
@@ -0,0 +1,421 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 The Foundry Visionmongers Ltd
+#include <algorithm>
+#include <cassert>
+#include <memory>
+
+#include <fmt/format.h>
+
+#include <openassetio/errors/exceptions.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/pluginSystem/HybridPluginSystemManagerImplementationFactory.hpp>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace pluginSystem {
+
+namespace {
+
+/**
+ * A ManagerInterface implementation that composes multiple child
+ * ManagerInterfaces and forwards API calls to one of the children.
+ *
+ * For most API methods, the call is forwarded to the first child that
+ * satisfies the required capability for that API call, or calls the
+ * base class implementation if no child satisfies the required
+ * capability. The priorirty order of children is defined by the order
+ * that they were provided to the constructor.
+ *
+ * For API calls that have no associated capability, either the first
+ * child is chosen, or the results from all children are merged - see
+ * method-specific docs for details.
+ */
+class HybridManagerInterface : public managerApi::ManagerInterface {
+  using ManagerInterfaces = std::vector<managerApi::ManagerInterfacePtr>;
+
+ public:
+  explicit HybridManagerInterface(ManagerInterfaces managerInterfacess)
+      : managerInterfaces_{std::move(managerInterfacess)} {
+    // Precondition.
+    assert(!managerInterfaces_.empty());
+
+    // Cache a mapping of the first child interface that supports each
+    // capability.
+    //
+    // Caching now avoids the need to call hasCapability() on
+    // the child implementations again (and repeatedly) in the API
+    // methods, which could be expensive if, for example, a network call
+    // is required, or we are calling out to Python and locking the
+    // GIL, etc. The disadvantage is that capabilities cannot change
+    // after plugins have been loaded.
+    managerInterfacesByCapability_.reserve(kCapabilityNames.size());
+    for (std::size_t capabilityIdx = 0; capabilityIdx < kCapabilityNames.size(); ++capabilityIdx) {
+      const auto capability = static_cast<Capability>(capabilityIdx);
+      for (const auto& managerInterface : managerInterfaces_) {
+        if (managerInterface->hasCapability(capability)) {
+          managerInterfacesByCapability_[capability] = managerInterface;
+          break;
```

Comment:
Ooo, that's a nice idea!  Though I think it would have to be configurable, or at least carefully designed - some failures might not be a good idea to retry.

Reply:
Though looking again - there's a bug!  We're not supposed to construct this cache until after `initialize(...)` has been called.  That's the point at which e.g. server connections are supposed to be made, not on construction. Construction is always supposed to be super lightweight.

Reply:
BTW not suggesting the fallback to other plugins if configured to do so, to be done here. But if you think it's worthwhile discussing further with the community, then please do log a ticket.

Reply:
Added https://github.com/OpenAssetIO/OpenAssetIO/issues/1374 to remember to discuss exception fallback behaviour.

Reply:
> We're not supposed to construct this cache until after initialize(...) has been called

Moved in c21eaef2